lib/Constants.sol

```diff
@@ -0,0 +1,50 @@
+// SPDX-License-Identifier: BUSL-1.1
+pragma solidity ^0.8.17;
+
+uint constant ONES = type(uint).max;
+uint constant TOPBIT = 1 << 255;
+
+// MIN_TICK and MAX_TICK should be inside the addressable range defined by the sizes of LEAF, LEVEL0, LEVEL1, LEVEL2
+int constant MIN_TICK = -524287;
+int constant MAX_TICK = -MIN_TICK;
+
+// sizes must match field sizes in structs.ts where relevant
+uint constant TICK_BITS = 24;
+uint constant OFFER_BITS = 32;
+
+// only power-of-two sizes are supported for LEAF_SIZE and LEVEL*_SIZE
+uint constant LEAF_SIZE_BITS = 2; 
+uint constant LEVEL0_SIZE_BITS = 6;
+uint constant LEVEL1_SIZE_BITS = 6;
+uint constant LEVEL2_SIZE_BITS = 6;
+
+int constant LEAF_SIZE = int(2 ** (LEAF_SIZE_BITS));
+int constant LEVEL0_SIZE = int(2 ** (LEVEL0_SIZE_BITS));
+int constant LEVEL1_SIZE = int(2 ** (LEVEL1_SIZE_BITS));
+int constant LEVEL2_SIZE = int(2 ** (LEVEL2_SIZE_BITS));
+
+uint constant LEAF_SIZE_MASK = ~(ONES << LEAF_SIZE_BITS);
+uint constant LEVEL0_SIZE_MASK = ~(ONES << LEVEL0_SIZE_BITS);
+uint constant LEVEL1_SIZE_MASK = ~(ONES << LEVEL1_SIZE_BITS);
+uint constant LEVEL2_SIZE_MASK = ~(ONES << LEVEL2_SIZE_BITS);
+
+int constant NUM_LEVEL1 = int(LEVEL2_SIZE);
+int constant NUM_LEVEL0 = NUM_LEVEL1 * LEVEL1_SIZE;
+int constant NUM_LEAFS = NUM_LEVEL0 * LEVEL0_SIZE;
+int constant NUM_TICKS = NUM_LEAFS * LEAF_SIZE;
+
+uint constant OFFER_MASK = ONES >> (256 - OFFER_BITS);
+
+
+
+// +/- 2**20-1 because only 20 bits are examined by the logPrice->price function
+int constant MIN_LOG_PRICE = -((1 << 20)-1);
+int constant MAX_LOG_PRICE = -MIN_LOG_PRICE;
```

Comment:
If understand things correctly, the limitation here is not only the 20 bits, but also that we assume we can represent `abs(logPrice)` in 20 bits, correct? Otherwise, `MIN_LOG_PRICE` could be `-(1 << 20)`.

Reply:
yes, the function never inspects logPrice if it's negative.

Reply:
Ok. I think it would be good to update the comment with this info